core/src/main/java/io/questdb/Telemetry.java

```diff
@@ -24,17 +24,133 @@
-    public static final short SYSTEM_EVENT_UP = 100;
-    public static final short SYSTEM_ILP_RESERVE_WRITER = 102;
+import io.questdb.cairo.*;
+import io.questdb.cairo.security.AllowAllCairoSecurityContext;
+import io.questdb.griffin.SqlCompiler;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.mp.MPSequence;
+import io.questdb.mp.QueueConsumer;
+import io.questdb.mp.RingQueue;
+import io.questdb.mp.SCSequence;
+import io.questdb.std.Misc;
+import io.questdb.std.ObjectFactory;
+import io.questdb.std.datetime.microtime.MicrosecondClock;
+import io.questdb.tasks.AbstractTelemetryTask;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.Closeable;
+
+public final class Telemetry<T extends AbstractTelemetryTask> implements Closeable {
+    private static final Log LOG = LogFactory.getLog(Telemetry.class);
+
+    private final boolean enabled;
+    private MicrosecondClock clock;
+    private MPSequence telemetryPubSeq;
+    private RingQueue<T> telemetryQueue;
+    private SCSequence telemetrySubSeq;
+    private TelemetryType<T> telemetryType;
+    private TableWriter writer;
+
+    private final QueueConsumer<T> taskConsumer = this::consume;
+
+    public Telemetry(TelemetryType<T> type, CairoConfiguration configuration) {
+        final TelemetryConfiguration telemetryConfiguration = type.getTelemetryConfiguration(configuration);
+        enabled = telemetryConfiguration.getEnabled();
+        if (enabled) {
+            clock = configuration.getMicrosecondClock();
+            telemetryType = type;
+            telemetryQueue = new RingQueue<>(type.getTaskFactory(), telemetryConfiguration.getQueueCapacity());
+            telemetryPubSeq = new MPSequence(telemetryQueue.getCycle());
+            telemetrySubSeq = new SCSequence();
+            telemetryPubSeq.then(telemetrySubSeq).then(telemetryPubSeq);
+        }
+    }
+
+    @Override
+    public void close() {
+        if (writer == null) {
+            return;
+        }
+
+        consumeAll();
+        Misc.free(telemetryQueue);
+
+        telemetryType.logStatus(writer, TelemetrySystemEvent.SYSTEM_DOWN, clock.getTicks());
+        writer = Misc.free(writer);
+    }
+
+    public void consume(T task) {
+        task.writeTo(writer, clock.getTicks());
+    }
+
+    public void consumeAll() {
+        if (enabled && telemetrySubSeq.consumeAll(telemetryQueue, taskConsumer)) {
+            writer.commit();
+        }
+    }
+
+    public void init(CairoEngine engine, SqlCompiler compiler, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        if (!enabled) {
+            return;
+        }
+        CharSequence sysPrefix = engine.getConfiguration().getSystemTableNamePrefix();
+        String tableName = sysPrefix + telemetryType.getTableName();
+
+        compiler.compile(telemetryType.getCreateSql(tableName), sqlExecutionContext);
```

Comment:
i think it would have been enough to pass the system prefix to getCreateSql() so it would look something like this:
```
@Override
public String getCreateSql(CharSequence prefix) {
    return "CREATE TABLE IF NOT EXISTS " + prefix + TABLE_NAME + " (" +
```
this way we do not need to call getTableName() before calling getCreateSql() .

Reply:
Replaced by a prefix argument. Still need to construct full `tableName` for the `tableToken`
Fixed in https://github.com/questdb/questdb/pull/2930/commits/889f293a98721949f07474684aaa0ecab02e28a6